File: src/host/runtime.ts
import { join } from 'node:path';
import { existsSync } from 'node:fs';
import { ChatGptBridgeStore } from './external/ChatGptBridgeStore';
import { appearanceFromSettings, type Appearance, type AxisKey } from '@shared/appearance';
import { SETTING_KEYS, sanitizeSetting, type HiddenMap } from '@shared/settings';
import type { WebviewHost } from '@shared/protocol';
import { AgentRegistry, type CustomAgentSetting } from './acp/AgentRegistry';
import { AccountManager } from './accounts/AccountManager';
import { AccountStore, FileVault } from './accounts/AccountStore';
import { DevinAccountProvider } from './accounts/devin';
import { LocalAccounts } from './accounts/local';
import { BridgeCore, type BridgeCoreOpts } from './bridgeCore';
import { msg } from './errors';
import { setHostLocale, t } from './i18n';
import type { HostPlatform, SettingsAffects } from './platform';
import { SessionManager } from './SessionManager';
import { SettingsCenter } from './settings';
import { acpiraHome, migrateOnce } from './store/dataDir';
import { TranscriptStore } from './store/TranscriptStore';

export interface HostRuntimeOpts {
  // ACPIRA_HOME / ~/.acpira by default; tests point it at a temp directory
  home?: string;
  chatgptBridgePath?: string;
}

// The one composition root: every host (the VS Code extension, the sidecar) builds the same registry, vault, account layer, session
// manager and settings center from a HostPlatform here, and reacts to its settings / focus events the same way. Views attach as
// BridgeCores and are torn down with the runtime
export class HostRuntime {
  readonly manager: SessionManager;
  readonly settings: SettingsCenter;
  readonly sessionsDir: string;
  private bridges = new Set<BridgeCore>();
  private activeRegistry: AgentRegistry;
  private unsubscribe: (() => void)[] = [];

  private constructor(private platform: HostPlatform, root: string, vault: FileVault, accountStore: AccountStore, opts: HostRuntimeOpts) {
    const log = (line: string) => platform.log(line);
    const registry = () => new AgentRegistry(this.read<Record<string, CustomAgentSetting>>('agents') ?? {});
    const runInTerminal = platform.runInTerminal.bind(platform);
    const toast = platform.toast.bind(platform);
    this.activeRegistry = registry();
    const accounts = new AccountManager({
      store: accountStore,
      providers: [new DevinAccountProvider(join(root, 'scratch'), () => this.activeRegistry.resolveBinary('devin'))],
      log, runInTerminal, toast,
    });

    this.sessionsDir = join(root, 'sessions');
    const bridgePath = opts.chatgptBridgePath ?? join(typeof __dirname === 'string' ? __dirname : join(process.cwd(), 'dist'), 'chatgpt-bridge.cjs');
    this.manager = new SessionManager({
      registry: this.activeRegistry,
      chatgpt: new ChatGptBridgeStore(join(root, 'bridges', 'chatgpt'), log, Date.now, existsSync(bridgePath) ? bridgePath : undefined),
      store: new TranscriptStore(this.sessionsDir, log, { onSaveError: (_id, error) => toast('error', t('host.saveFailed', { error })) }),
      log,
      cwd: () => platform.cwd(),
      defaultAgent: () => this.read<string>('defaultAgent') ?? 'grok',
      runInTerminal, toast, accounts,
      localAccounts: new LocalAccounts({ env: agent => ({ ...process.env, ...this.activeRegistry.get(agent).env }) }),
      compaction: () => ({ atTokens: this.read<number>('compactAtTokens') ?? 300_000, auto: this.read<boolean>('autoCompact') ?? true }),
      hidden: () => this.read<HiddenMap>('hiddenOptions') ?? {},
      scope: () => sanitizeSetting('sessionScope', this.read('sessionScope')),
    });

    // The settings page's backend: reads / writes acpira.*, scans agent inventories; every bridge gets a subscription
    this.settings = new SettingsCenter({
      read: key => platform.readSetting(key),
      write: (key, value) => platform.writeSetting(key, value),
      writeAppearance: (axis, value) => platform.writeSetting(`appearance.${axis}`, value),
      hostLanguage: () => platform.hostLanguage(),
      registry: () => this.manager.registry,
      runtimeInfo: agent => this.manager.runtimeInfo(agent),
      home: () => platform.home(),
      cwd: () => platform.cwd(),
    });

    this.unsubscribe.push(
      platform.onSettingsChanged(affects => this.settingsChanged(affects)),
      // Coming back from an external terminal where a CLI was installed or removed: re-check the executables right away; coming back from
      // another window (VS Code, Cursor or IDEA, same ~/.acpira): pick up the sessions and accounts it created or deleted
      platform.onWindowFocus(() => {
        void this.manager.reprobe();
        void this.manager.refreshIndex();
        accounts.reload().catch(e => log(`account reload failed: ${msg(e)}`));
      }),
    );
  }

  static async create(platform: HostPlatform, opts: HostRuntimeOpts = {}): Promise<HostRuntime> {
    const log = (line: string) => platform.log(line);
    // ~/.acpira (ACPIRA_HOME) holds accounts.json, secrets.json, sessions/, scratch/; a platform with a legacy tree has it copied once
    const root = opts.home ?? acpiraHome();
    const vault = new FileVault(join(root, 'secrets.json'), log);
    if (platform.legacy) await migrateOnce({ from: platform.legacy.from, to: root, oldVault: platform.legacy.vault, newVault: vault, log });
    const accountStore = new AccountStore(join(root, 'accounts.json'), vault, log);
    await accountStore.load();
    const runtime = new HostRuntime(platform, root, vault, accountStore, opts);
    await runtime.manager.init();
    setHostLocale(runtime.settings.locale());
    return runtime;
  }

  private read<T>(key: string): T | undefined {
    return this.platform.readSetting(key) as T | undefined;
  }

  appearance(): Appearance {
    return appearanceFromSettings((k: AxisKey) => this.platform.readSetting(`appearance.${k}`));
  }

  // One BridgeCore per view; the caller owns the transport and calls dispose() on the returned core when the view goes away
  attachView(opts: BridgeCoreOpts & { host: WebviewHost }): BridgeCore {
    const core = new BridgeCore({ manager: this.manager, settings: this.settings, platform: this.platform, appearance: () => this.appearance() }, opts);
    this.bridges.add(core);
    return core;
  }

  detachView(core: BridgeCore) {
    if (this.bridges.delete(core)) core.dispose();
  }

  private settingsChanged(affects: SettingsAffects) {
    if (affects('appearance')) for (const b of this.bridges) b.pushAppearance();
    if (affects('agents')) { this.activeRegistry = new AgentRegistry(this.read<Record<string, CustomAgentSetting>>('agents') ?? {}); this.manager.setRegistry(this.activeRegistry); }
    if (affects('hiddenOptions')) this.manager.emitHidden();
    // Any knob the settings page shows (language, defaultAgent, compaction, …): re-push the view and follow a language change host-side.
    // Checked per key, not as "anything but appearance / agents": one shell event may carry an appearance axis and a language change together
    if (SETTING_KEYS.some(k => affects(k))) {
      this.settings.emit();
      setHostLocale(this.settings.locale());
    }
  }

  async dispose() {
    for (const u of this.unsubscribe) u();
    this.unsubscribe = [];
    for (const b of this.bridges) b.dispose();
    this.bridges.clear();
    await this.manager.dispose().catch(e => this.platform.log(`dispose failed: ${msg(e)}`));
  }
}

export const createHostRuntime = HostRuntime.create;

File: rust/crates/acpira-host/src/runtime.rs
//! The one composition root (mirror of src/host/runtime.ts): registry, vault, account layer, session manager and
//! settings center built from a platform, reacting to its settings / focus events. Views attach as BridgeCores

use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::Arc;

use anyhow::Result;
use serde_json::Value;

use acpira_shared::agent_order::AgentPrefs;
use acpira_shared::appearance::{Appearance, appearance_from_settings};
use acpira_shared::protocol::WebviewHost;
use acpira_shared::settings::{SETTING_KEYS, id_list, is_hidden_map};
use acpira_shared::sidecar::InitialView;

use crate::accounts::account_manager::AccountManager;
use crate::accounts::account_store::{AccountStore, FileVault};
use crate::accounts::devin::DevinAccountProvider;
use crate::accounts::local::LocalAccounts;
use crate::acp::agent_registry::AgentRegistry;
use crate::acp::session::CompactionPolicy;
use crate::bridge_core::{BridgeCore, Post};
use crate::external::chatgpt_store::ChatGptBridgeStore;
use crate::i18n::{set_host_locale, tp};
use crate::session_manager::{ManagerDeps, SessionManager};
use crate::settings::{SettingsCenter, SettingsDeps};
use crate::sidecar::platform::{Affects, SidecarPlatform};
use crate::store::transcript_store::{LogFn, TranscriptStore};

pub struct HostRuntime {
  pub manager: Arc<SessionManager>,
  pub settings: Arc<SettingsCenter>,
  pub sessions_dir: PathBuf,
  platform: Arc<SidecarPlatform>,
  accounts: Arc<AccountManager>,
  bridges: parking_lot::Mutex<Vec<Arc<BridgeCore>>>,
}

impl HostRuntime {
  /// `root`: ACPIRA_HOME / ~/.acpira; `bridge_exe`: the executable the ChatGPT connection prompt names
  pub async fn create(platform: Arc<SidecarPlatform>, root: PathBuf, bridge_exe: Option<String>) -> Result<Arc<HostRuntime>> {
    let p = platform.clone();
    let log: LogFn = Arc::new(move |line: &str| p.log(line));
    let vault = Arc::new(FileVault::new(root.join("secrets.json"), log.clone()));
    let account_store = Arc::new(AccountStore::new(root.join("accounts.json"), vault, log.clone()));
    account_store.load().await?;

    let read = {
      let p = platform.clone();
      move |key: &str| p.read_setting(key)
    };
    let registry = Arc::new(AgentRegistry::new(&read("agents").unwrap_or(Value::Null)));

    // The manager is created below; the Devin provider resolves its binary through whatever registry is current then
    let mgr_slot: Arc<parking_lot::Mutex<Option<std::sync::Weak<SessionManager>>>> = Default::default();
    let slot = mgr_slot.clone();
    let fallback = registry.clone();
    let devin = DevinAccountProvider::new(
      root.join("scratch"),
      Arc::new(move || {
        let reg = slot.lock().as_ref().and_then(|w| w.upgrade()).map(|m| m.registry()).unwrap_or_else(|| fallback.clone());
        Box::pin(async move { reg.resolve_binary("devin").await })
      }),
    );
    let terminal = {
      let p = platform.clone();
      Arc::new(move |title, command, args, env| p.run_in_terminal(title, command, args, env))
    };
    let toast = {
      let p = platform.clone();
      Arc::new(move |level: &str, text: &str| p.toast(level, text))
    };
    let accounts = AccountManager::new(account_store, vec![Arc::new(devin)], log.clone(), terminal.clone(), toast.clone());

    let sessions_dir = root.join("sessions");
    let save_toast = toast.clone();
    let store = TranscriptStore::new(
      sessions_dir.clone(),
      log.clone(),
      Some(Arc::new(move |_id: &str, error: &str| save_toast("error", &tp("host.saveFailed", &[("error", error)])))),
    );
    let chatgpt = ChatGptBridgeStore::new(root.join("bridges").join("chatgpt"), log.clone(), bridge_exe);

    let local_env_slot = mgr_slot.clone();
    let local_accounts = LocalAccounts::new(Arc::new(move |agent: &str| {
      let mut env: HashMap<String, String> = std::env::vars().collect();
      if let Some(m) = local_env_slot.lock().as_ref().and_then(|w| w.upgrade())
        && let Some(def) = m.registry().try_get(agent)
      {
        env.extend(def.env.clone().unwrap_or_default());
      }
      env
    }));

    let r = |p: &Arc<SidecarPlatform>| {
      let p = p.clone();
      move |k: &str| p.read_setting(k)
    };
    let (r1, r2, r3, r4, r5) = (r(&platform), r(&platform), r(&platform), r(&platform), r(&platform));
    let cwd_p = platform.clone();
    let manager = SessionManager::new(
      registry,
      ManagerDeps {
        store,
        chatgpt: Some(chatgpt),
        log: log.clone(),
        cwd: Arc::new(move || cwd_p.cwd()),
        default_agent: Arc::new(move || r1("defaultAgent").and_then(|v| v.as_str().map(str::to_owned)).unwrap_or_else(|| "grok".into())),
        agent_prefs: Arc::new(move || AgentPrefs {
          order: id_list(&r2("agentOrder").unwrap_or(Value::Null)),
          disabled: id_list(&r2("disabledAgents").unwrap_or(Value::Null)),
        }),
        run_in_terminal: terminal,
        toast,
        accounts: Some(accounts.clone()),
        local_accounts: Some(local_accounts),
        compaction: Arc::new(move || CompactionPolicy {
          at_tokens: r3("compactAtTokens").and_then(|v| v.as_f64()).unwrap_or(300_000.0),
          auto: r3("autoCompact").and_then(|v| v.as_bool()).unwrap_or(true),
        }),
        hidden: Arc::new(move || {
          r4("hiddenOptions").filter(is_hidden_map).and_then(|v| serde_json::from_value(v).ok()).unwrap_or_default()
        }),
        scope: Arc::new(move || match r5("sessionScope").and_then(|v| v.as_str().map(str::to_owned)).as_deref() {
          Some("all") => "all".into(),
          _ => "workspace".into(),
        }),
      },
    );
    *mgr_slot.lock() = Some(Arc::downgrade(&manager));

    let (m1, m2, m3) = (manager.clone(), manager.clone(), manager.clone());
    let (pr, pw, ph, phome, pcwd) = (platform.clone(), platform.clone(), platform.clone(), platform.clone(), platform.clone());
    let settings = Arc::new(SettingsCenter::new(SettingsDeps {
      read: Arc::new(move |k| pr.read_setting(k)),
      write: Arc::new(move |k, v| {
        let p = pw.clone();
        Box::pin(async move { p.write_setting(&k, v).await })
      }),
      host_language: Arc::new(move || ph.host_language()),
      registry: Arc::new(move || m1.registry()),
      runtime_info: Arc::new(move |a| m2.runtime_info(a)),
      health: Arc::new(move |a| m3.agent_health(a)),
      home: Arc::new(move || phome.home()),
      cwd: Arc::new(move || pcwd.cwd()),
    }));

    let runtime =
      Arc::new(HostRuntime { manager, settings, sessions_dir, platform: platform.clone(), accounts, bridges: Default::default() });
    let weak = Arc::downgrade(&runtime);
    platform.on_settings_changed(Arc::new(move |affects| {
      if let Some(rt) = weak.upgrade() {
        rt.settings_changed(affects);
      }
    }));
    let weak = Arc::downgrade(&runtime);
    platform.on_window_focus(Arc::new(move || {
      let Some(rt) = weak.upgrade() else { return };
      // Back from a terminal where a CLI was installed, or from another window sharing ~/.acpira
      tokio::spawn(async move {
        rt.manager.reprobe().await;
        rt.manager.refresh_index().await;
        rt.accounts.reload().await;
      });
    }));
    runtime.manager.init().await;
    set_host_locale(runtime.settings.locale());
    Ok(runtime)
  }

  pub fn appearance(&self) -> Appearance {
    appearance_from_settings(|k| self.platform.read_setting(&format!("appearance.{k}")))
  }

  pub fn attach_view(
    self: &Arc<Self>,
    host: WebviewHost,
    initial: Option<InitialView>,
    blob_base: Option<String>,
    post: Post,
  ) -> Arc<BridgeCore> {
    let rt = Arc::downgrade(self);
    let appearance = Arc::new(move || rt.upgrade().map(|r| r.appearance()).unwrap_or_default());
    let core =
      BridgeCore::new(self.manager.clone(), self.settings.clone(), self.platform.clone(), appearance, post, host, blob_base, initial);
    self.bridges.lock().push(core.clone());
    core
  }

  pub fn detach_view(&self, core: &Arc<BridgeCore>) {
    let mut bridges = self.bridges.lock();
    if let Some(i) = bridges.iter().position(|b| Arc::ptr_eq(b, core)) {
      bridges.remove(i);
      drop(bridges);
      core.dispose();
    }
  }

  fn settings_changed(self: &Arc<Self>, affects: Affects) {
    if affects(Some("appearance")) {
      for b in self.bridges.lock().clone() {
        b.push_appearance();
      }
    }
    if affects(Some("agents")) {
      let custom = self.platform.read_setting("agents").unwrap_or(Value::Null);
      self.manager.set_registry(Arc::new(AgentRegistry::new(&custom)));
    }
    if affects(Some("hiddenOptions")) {
      self.manager.emit_hidden();
    }
    if affects(Some("agentOrder")) || affects(Some("disabledAgents")) {
      self.manager.emit_agents();
    }
    // Checked per key: one event may carry an appearance axis and a language change together
    if SETTING_KEYS.iter().any(|k| affects(Some(k))) {
      self.settings.emit();
      set_host_locale(self.settings.locale());
    }
  }

  pub async fn dispose(&self) {
    let bridges: Vec<_> = self.bridges.lock().drain(..).collect();
    for b in bridges {
      b.dispose();
    }
    self.manager.dispose().await;
  }
}
